sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsLogger.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.metrics;
+
+import java.io.Serializable;
+import java.util.Date;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+import org.slf4j.Logger;
+
+@Experimental(Kind.METRICS)
+public interface MetricsLogger extends Serializable {
+  Lock REPORTING_LOCK = new ReentrantLock();
+  AtomicLong LAST_REPORTED_MILLIS = new AtomicLong(System.currentTimeMillis());
+
+  default void tryLoggingMetrics(
+      String header, String namespace, long minimumLoggingFrequencyMillis, boolean resetMetrics) {
+    if (REPORTING_LOCK.tryLock()) {
```

Comment:
Is it possible to use this style for logging, if it exists? If not ignore this suggestion.


        try (Autolock alock = new Autolock(lock)) {
            // Whatever you need to do while you own the lock
        } //unlocks here
        
        See:

https://stackoverflow.com/questions/151917/autolock-in-java-how-to

Reply:
`Lock` is not auto closeable by default and IMHO `tryLock` doesn't fit well with try-with-resources style (it hides the semantics of `tryLock`).